### ISSUE TITLE
Crash during disconnection / fallback attempts

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-cocoa#901]
### Which version of the Ably SDK are you using? 
`1.0.10`

### On which platform does the issue happen? 
`iOS11` (waiting for confirmation from customers)

### Are you using Carthage? 
Unknown, waiting on customer to confirm.

### Are you using Cocoapods? 
Unknown, waiting on customer to confirm.

### Which version of Xcode are you using? 
Unknown, waiting on customer to confirm.

### What did you do?
A customer has an ATM configuration which was deployed this week.  However, we discovered that they had omitted to configure the `environment` correctly, so only had fallbacks configured correctly.

Our strategy today was to force traffic across to the new endpoints (clusters) by serving a 50x on the global cluster (used as no env set), which in turn should have automatically resulted in the lib using a fallback host.

This however was further compounded by a problem whereby the fallback hosts had an incorrect certificate, due to a typo in the original fallback hosts being used (thus causing a certificate mismatch).

### What did you expect to happen?
We expected the clients to remain in a disconnected / retry / suspended state cycle due to 50x on the primary endpoints, and cert issues on the fallback hosts.

### What happened instead?

Lots of devices crashed.  Specifically `Crashed: io.ably.main EXC_BAD_ACCESS KERN_INVALID_ADDRESS` errors and `Fatal Exception: NSGenericException Collection ARTChannelsNosyncIterable was mutated whilst being enumerated`

<img width="1046" alt="螢幕快照 2019-10-02 上午1 40 51" src="https://user-images.githubusercontent.com/43789/65986717-5b477c80-e47c-11e9-9ff9-964d663e056e.png">

<img width="1046" alt="螢幕快照 2019-10-02 上午1 40 42" src="https://user-images.githubusercontent.com/43789/65986718-5be01300-e47c-11e9-8992-5787f948f89d.png">

### Logs

We are asking the customer to provide more logs.  We will create a support ticket so that any sensitive information is not included in this issue.

**Comment by ricardopereira** [comment]
@mattheworiordan It seems to be the same issue as #879. There's already a tentative fix in beta: https://github.com/ably/ably-cocoa/releases/tag/1.1.11-beta.1.

**Comment by tcard** [comment]
> Lots of devices crashed. Specifically `Crashed: io.ably.main EXC_BAD_ACCESS KERN_INVALID_ADDRESS` errors 

This one does indeed looks like #879.

> and `Fatal Exception: NSGenericException Collection ARTChannelsNosyncIterable was mutated whilst being enumerated`

This stack trace is different, but the underlying issue looks the same: a release on the client's dispatch queue causes ARTChannels's internal NSDictionary to be deallocated while it's being used by the library's dispatch queue. There's no other way that mutation of the NSDictionary could happen at that point.

Both issues _should theoretically_ be fixed by #879. However, we haven't been able to reproduce anything like that; it involves concurrency and data races and thus isn't deterministic. I've in particular tried disconnect/retry/suspend cycles as you described, with no luck.

It would be of great help if we could get more data on the exact environment the crashes are happening, and if there's some particular factor that increases or determines its chances of happening: iOS version, device version, etc.

**Comment by tcard** [comment]
Update:

The first trace does look like what #819 fixed. Again, we don't have _proof_ it's fixed, but it should be.

The second looks like a duplicate of #918, which is fixed.

I'm closing this issue since both are supposed to be fixed as of 1.1.13.
